File: week_1/day-5-string-hashing/valid_anagram.py
# Input
s = "listen"
t = "silent"

def isAnagram(s, t):
    if len(s) != len(t):
        return False

    freq = {}

    # Count frequency of s
    for ch in s:
        freq[ch] = freq.get(ch, 0) + 1

    # Decrease frequency using t
    for ch in t:
        if ch not in freq or freq[ch] == 0:
            return False
        freq[ch] -= 1

    return True

print(isAnagram(s, t))  # Output: True

File: week_1/day-7-revision/validAnagram.js
// Input
const s = "listen";
const t = "silent";

function isAnagram(s, t) {
    if (s.length !== t.length) return false;

    const freq = {};

    // Count frequency of s
    for (let ch of s) {
        freq[ch] = (freq[ch] || 0) + 1;
    }

    // Decrease frequency using t
    for (let ch of t) {
        if (!freq[ch]) return false;
        freq[ch]--;
    }

    return true;
}

console.log(isAnagram(s, t)); // Output: true
